src/dev/pi4ioe5v6408.rs

```diff
@@ -0,0 +1,337 @@
+    /// - `default_pull_mode` - The default pull mode for all pins (The pins are configured as inputs on reset)
+    pub fn new(i2c: I2C, addr: bool, default_pull_mode: PullMode) -> Result<Self, I2C::BusError> {
+        Self::with_mutex(i2c, addr, default_pull_mode)
+    }
+}
+
+impl<I2C, M> Pi4ioe5v6408<M>
+where
+    I2C: crate::I2cBus,
+    M: shared_bus::BusMutex<Bus = Driver<I2C>>,
+{
+    /// Create a new driver for the `PI4IOE5V6408` "Low-voltage Translating 8-bit I2C-bus I/O Expander"
+    /// with a mutex
+    ///
+    /// # Arguments
+    /// - `i2c` - The I2C bus the device is connected to
+    /// - `addr` - The address of the device. The address is 0x43 if `addr` is `false` and 0x44 if `addr` is `true`
+    /// - `default_pull_mode` - The default pull mode for all pins (The pins are configured as inputs on reset)
+    pub fn with_mutex(
+        i2c: I2C,
+        addr: bool,
+        default_pull_mode: PullMode,
+    ) -> Result<Self, I2C::BusError> {
+        Ok(Self(shared_bus::BusMutex::create(Driver::new(
+            i2c,
+            addr,
+            default_pull_mode,
+        )?)))
+    }
+
+    pub fn split<'a>(&'a mut self) -> Parts<'a, I2C, M> {
+        Parts {
+            io0: crate::Pin::new(0, &self.0),
+            io1: crate::Pin::new(1, &self.0),
+            io2: crate::Pin::new(2, &self.0),
+            io3: crate::Pin::new(3, &self.0),
+            io4: crate::Pin::new(4, &self.0),
+            io5: crate::Pin::new(5, &self.0),
+            io6: crate::Pin::new(6, &self.0),
+            io7: crate::Pin::new(7, &self.0),
+        }
+    }
+}
+
+pub struct Parts<'a, I2C, M = shared_bus::NullMutex<Driver<I2C>>>
+where
+    I2C: crate::I2cBus,
+    M: shared_bus::BusMutex<Bus = Driver<I2C>>,
+{
+    pub io0: crate::Pin<'a, crate::mode::Input, M>,
+    pub io1: crate::Pin<'a, crate::mode::Input, M>,
+    pub io2: crate::Pin<'a, crate::mode::Input, M>,
+    pub io3: crate::Pin<'a, crate::mode::Input, M>,
+    pub io4: crate::Pin<'a, crate::mode::Input, M>,
+    pub io5: crate::Pin<'a, crate::mode::Input, M>,
+    pub io6: crate::Pin<'a, crate::mode::Input, M>,
+    pub io7: crate::Pin<'a, crate::mode::Input, M>,
+}
+
+#[allow(dead_code)]
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+enum Regs {
+    DeviceIdControl = 0x01,
+    IODirection = 0x03,
+    OutputPort = 0x05,
+    OutputHighImpedance = 0x07,
+    InputDefaultState = 0x09,
+    PullUpPullDownEnable = 0x0b,
+    PullUpPullDownSelection = 0x0d,
+    InputStatusRegister = 0x0f,
+    InterruptMaskRegister = 0x11,
+    InterruptStatusRegister = 0x13,
+}
+
+impl From<Regs> for u8 {
+    fn from(r: Regs) -> u8 {
+        r as u8
+    }
+}
+
+/// The pull mode for an input pin
+pub enum PullMode {
+    /// Pull the pin up
+    PullUp,
+    /// Pull the pin down
+    PullDown,
+    /// Don't activate pull-up or pull-down resistors and leave the pin floating
+    Floating,
+}
```

Comment:
I don't think this "default pull mode" is needed when we get proper per-pin pull-up/down configuration support with your other changes?

Reply:
Before I rebase/refactor:
The rationale behind having to specify this, is that per default the pins are configured as pull down, which can be very misleading. E.g. when I call into_input() I would expect the pin to be floating or with pullup. 

The nicest solution would probably to adapt the traits, so that they include the default pull up/down/floating mode of the pin, similar to here: https://docs.rs/esp32c6-hal/latest/esp32c6_hal/gpio/struct.GpioPin.html
But this is a bigger refactoring, that I dont currently have the time to tackle.

Please advise on how to proceed here

Reply:
I'd argue the default should always be floating inputs.  That's how most I/O pins and port expanders work, as that's electrically the only state without possible side effects.

Reply:
I agree that floating is a suitable default.

But I still need a way to specify that the configured pull mode should not be changed. The reason is, that I put the MCU into deepsleep and wake up when the port expander signals an interrupt. In that case I dont want the input pins to be reconfigured as floating, actually I dont want any reconfiguration at all. In the updated PR I have added a `no_initialize` flag for that. Is this an acceptable compromise?

(Follow-up PR with pull-up/pull-down traits is ready locally).

Reply:
I like the approach of providing the option to retain previous configuration.  Although you have to keep in mind that this means the type-state of the `Pin` objects may be wrong after doing that.